ostd/src/cpu/local/dyn_cpu_local.rs

```diff
@@ -0,0 +1,238 @@
+
+use bitvec::prelude::{bitvec, BitVec};
+
+use super::{AnyStorage, CpuLocal};
+use crate::{
+    cpu::{all_cpus, num_cpus, CpuId, PinCurrentCpu},
+    mm::{paddr_to_vaddr, FrameAllocOptions, Segment, Vaddr, PAGE_SIZE},
+    trap::DisabledLocalIrqGuard,
+    Result,
+};
+
+/// A dynamically-allocated storage for a CPU-local variable of type `T`.
+///
+/// Such a CPU-local storage should be allocated and deallocated by
+/// [`DynCpuLocalChunk`], not directly. Dropping it without deallocation
+/// will cause panic.
+///
+/// When dropping a `CpuLocal<T, DynamicStorage<T>>`, we have no way to know
+/// which `DynCpuLocalChunk` the CPU-local object was originally allocated
+/// from. Therefore, we rely on the user to correctly manage the corresponding
+/// `DynCpuLocalChunk`, ensuring that both allocation and deallocation of
+/// `CpuLocal<T, DynamicStorage<T>>` occur within the same chunk.
+///
+/// To properly deallocate the CPU-local object, the user must explicitly call
+/// the appropriate `DynCpuLocalChunk`'s `try_dealloc<T>()`. Otherwise,
+/// dropping it directly will cause a panic.
+pub struct DynamicStorage<T>(*mut T);
+
+unsafe impl<T> AnyStorage<T> for DynamicStorage<T> {
+    fn get_ptr_on_current(&self, guard: &DisabledLocalIrqGuard) -> *const T {
+        self.get_ptr_on_target(guard.current_cpu())
+    }
+
+    fn get_ptr_on_target(&self, cpu_id: CpuId) -> *const T {
+        let bsp_va = self.0 as usize;
+        let va = bsp_va + cpu_id.as_usize() * CHUNK_SIZE;
+        va as *mut T
+    }
+
+    fn get_mut_ptr_on_target(&mut self, cpu: CpuId) -> *mut T {
+        self.get_ptr_on_target(cpu).cast_mut()
+    }
+}
+
+impl<T> Drop for DynamicStorage<T> {
+    fn drop(&mut self) {
+        panic!(
+            "Do not drop `DynamicStorage<T>` directly. \
+            Use `DynCpuLocalChunk::try_dealloc<T>` instead."
+        );
+    }
+}
+
+impl<T: Sync + alloc::fmt::Debug + 'static> alloc::fmt::Debug for CpuLocal<T, DynamicStorage<T>> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let mut list = f.debug_list();
+        for cpu in all_cpus() {
+            let val = self.get_on_cpu(cpu);
+            list.entry(&(&cpu, val));
+        }
+        list.finish()
+    }
+}
+
+impl<T> CpuLocal<T, DynamicStorage<T>> {
+    /// Creates a new dynamically-allocated CPU-local object, and
+    /// initializes it with `init_values`.
+    ///
+    /// The given `ptr` points to the variable located on the BSP.
+    ///
+    /// Please do not call this function directly. Instead, use
+    /// `DynCpuLocalChunk::alloc`.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the new per-CPU object belongs to an
+    /// existing [`DynCpuLocalChunk`], and does not overlap with any existing
+    /// CPU-local object.
+    unsafe fn __new_dynamic(ptr: *mut T, init_values: &mut impl FnMut(CpuId) -> T) -> Self {
+        let mut storage = DynamicStorage(ptr);
```

Comment:
```suggestion
        let mut storage = DynamicStorage(NonNull::new(ptr).unwrap());
```

Using `ptr: *mut T` here will cause `CpuLocalBox` to have more than 8 bytes, i.e.,
 - `Option<NonNull<u8>>` needs exactly 8 bytes but
 - `Option<*mut u8>` will need more than 8 bytes.

So I suggest to consider to use `NonNull`  as `ptr` must not be null.